GSoC '25 project idea proposal (#258)
pybamm-team/pybamm.org: content/gsoc/2025/_index.md

<hr>

> ## Model API Dispatching

Reviewer: ```suggestion
## Adding a dispatching mechanism for third-party models
```

> With a constantly expanding user base and community of PyBaMM, including and working with third-party battery models becomes a hassle for PyBaMM users if they have less technical knowledge about managing and configuring Python development environments. This project aims to develop a standardized framework for the integration, distribution, and dynamic utilization of third-party battery models implemented with PyBaMM, addressing current limitations in accessibility for users with varying technical backgrounds.

Reviewer: ```suggestion
With a constantly expanding user base and community, including and working with third-party battery models has become a hassle for PyBaMM users. This project aims to develop a standardized framework for the integration, distribution, and dynamic utilization of third-party battery models implemented with PyBaMM, addressing current limitations in accessibility for users with varying technical backgrounds.
```

> This project is an extension of the [cookiecutter project (GSoC 2024)](https://github.com/pybamm-team/cookiecutter) that aimed to reduce the entry barrier among scientists and researchers who want to use PyBaMM and distribute their work to other community members. Extending that work, this year through this project we aim to establish a standardized framework for model distribution and dispatching, ensuring seamless integration with PyBaMM. It will define guidelines for structuring models and documentation to facilitate easy distribution and dynamic loading via model [entry points](https://packaging.python.org/en/latest/specifications/entry-points/) through the dispatching API.

Reviewer: ```suggestion
This project serves as an extension of the [cookiecutter project (GSoC 2024)](https://github.com/pybamm-team/pybamm-cookie), which aimed to reduce the entry barrier for building and distributing PyBaMM projects with the community. This project will build on top of the 2024 project, aiming to establish a standardized model distribution and dispatching framework with seamless PyBaMM integration. The project will define guidelines for structuring models and documentation to facilitate easy distribution and dynamic loading of models. The models will be distributed via model [entry points](https://packaging.python.org/en/latest/specifications/entry-points/) through a new dispatching API.
```

> This project is an extension of the [cookiecutter project (GSoC 2024)](https://github.com/pybamm-team/cookiecutter) that aimed to reduce the entry barrier among scientists and researchers who want to use PyBaMM and distribute their work to other community members. Extending that work, this year through this project we aim to establish a standardized framework for model distribution and dispatching, ensuring seamless integration with PyBaMM. It will define guidelines for structuring models and documentation to facilitate easy distribution and dynamic loading via model [entry points](https://packaging.python.org/en/latest/specifications/entry-points/) through the dispatching API.

Author: Yes, I was planning on adding additional context and reference links for this. Thank you for attaching some of those links!

> Enhancements to the [existing copier template](https://github.com/pybamm-team/pybamm-cookie) to enforce standardized plugin structures for third-party models will also be a parallel primary goal of this project. The packaging and distribution system will be refactored in the copier template in compliance with model entry points/dispatch API within PyBaMM. To streamline model retrieval, a simple PyBaMM CLI interface/script will enable users to pull models effortlessly from the indexing repository incorporating a serverless indexing mechanism for pulling third-party models, complete with versioning support to maintain compatibility with specific PyBaMM versions. Additionally, the project will also implement caching and lazy loading mechanisms to optimize performance as a stretch goal.

Reviewer: ```suggestion

The mentee will enhance the [existing copier template](https://github.com/pybamm-team/pybamm-cookie) to enforce standardized plugin structures for third-party models, which will also be a parallel primary goal of this project. The packaging and distribution system of the template will be refactored to comply with the new model entry points/dispatch API of PyBaMM. To streamline model retrieval, a simple PyBaMM CLI interface/script will enable users to pull models effortlessly from the indexing repository, incorporating a serverless indexing mechanism for pulling third-party models. The project will also implement caching and lazy loading mechanisms to optimize performance as a stretch goal.
```

> The student will receive an opportunity to not only perform and learn an aspect of software engineering tasks with Python, including adding features, writing tests, writing user-facing documentation, usage examples, CI/CD pipelines for testing automation and deployment, and so on, but will also a significant opensource contribution in scientific computing and battery modelling research.

Reviewer: ```suggestion
The student will have an opportunity to learn software engineering with Python, including adding features, writing tests, writing user-facing documentation, designing examples, and building CI/CD pipelines. Additionally, their work will be a significant open-source contribution to scientific computing and battery modeling research.
```

### Expected Outcomes

> - A dispatch API and a framework that can be used to integrate, package, and distribute among the PyBaMM users. An effective user experience for using third-party PyBaMM models.

Reviewer: ```suggestion
- A dispatch API and a framework for integrating, packaging, and distributing battery models amongst PyBaMM users.
- An effective user experience for using third-party PyBaMM models.
```

> - A dispatch API and a framework that can be used to integrate, package, and distribute among the PyBaMM users. An effective user experience for using third-party PyBaMM models.

Reviewer: We should differentiate the changes required in PyBaMM (dispatch mechanism and model entry points) and pybamm-cookie (model templates) better. You can also link to your previous efforts of doing these things.

Author: Sure, I can give context with my previous work as a PoC, maybe even link the PRs and commits, but the implementation part would be fairly new to the student themselves.

> - A dispatch API and a framework that can be used to integrate, package, and distribute among the PyBaMM users. An effective user experience for using third-party PyBaMM models.
> - An effective guide and documentation on how to use the template to make your model and distribute it.
> - A PyBaMM-based CLI to install third-party models and parameter sets on demand.

Reviewer 2: ```suggestion
- A dispatch API and a framework that can be used to integrate, package, and distribute third-party PyBaMM models.
- A practical guide and documentation on how to use the template to add and distribute one's own model(s)
- A PyBaMM-based CLI to install third-party models and parameter sets on demand.
```

That said, I'm not sure if a CLI is going to be the best way to install external models or parameter sets. A more programmatic experience that can be loosely based on [`fsspec`](https://filesystem-spec.readthedocs.io/en/latest/features.html) would be much nicer. Something like

```python
import pybamm
pybamm.models.Model("git://path/to/myrepo.git")
```

or something like

```python
import pybamm
pybamm.models.Model("pypi::my_published_model_on_pypi")
```

which would provide an entry point for PyBaMM that the dispatching API can hook into, and also support dependency constraints – say, the model being able to run only with PyBaMM 25.1 or newer, as we will have metadata to go through for this.

(the public API can be discussed later, of course).

Note that the complexity rises as people can also add their own parameter sets with the template. For example, an external parameter set published by a research lab (not validated and put into the PyBaMM repo) which contains no SEI-related parameters can't be expected to work with a model with SEI enabled. We need a way to support such constraints and raise helpful errors.

These are all aspects that we can expect a "good" student to consider with their proposal, but my emphasis is that we should ourselves provide a few starter ideas at least, that they can base off their own with our initial directions – and they should be more carefully mentioned here so that we don't mislead a prospective student in the sense that their proposal deviates highly from our proposed outcome.

Author: I agree with your approach, it would be better to have registries in a JSON format, lets say, which would contain the constraints and scope of a model. With the API, I think it would be better to map the `pypi` and `git` URLs with an alias within the JSON. A centralised registry is easier to manage and update as we don't have to go and refactor the code everytime for edge cases. What I mean by that is what I previously worked on, you can initialise a model using `pybamm.models.Model("author/year")`, it would search the default entry points in the PyBaMM repository first, then it would go to the local cache and in the last case fetch it using http and load it lazily as we do not want to bloat the main PyBaMM code with third party models.

Author: We can also remove the CLI part if it is not needed, it is just a thoughtful feature that I added while writing my PoC, and this comment made it clear.

Reviewer 2: On second thought, an HTTP-based interface (or a centralised API) is another task that deserves more discussion and could be another stretch goal or GSoC project. We should limit ourselves and say that the way to load a custom model is to install PyBaMM, install a package from PyPI or any other index (or a local package) into the environment that contains the custom model, and hook into entry points with PyBaMM-side logic. This is, unless you have better ideas, of course.

Using the model would always be handled lazily as the entry point wouldn't be globally managed.

I think adding some details about how the models are supposed to look like would suffice here. We should remove mentions of the CLI for now, and keep the section intentionally vague – and at the same time, include a few pointers.

Author: @agriyakhetarpal I have made modifications since our last discussion on this thread which somewhat resembles to what you have mentioned, I have removed the CLI mentions, I think the way we could limit ourselves from loading the models is if we just contain everything under the same function `pybamm.models.Model()`, even the API that fetches the model from the registry, unless of course there is some kind of discrete functionality that occurs. This would make it simple to fetch from a user point of view.

Reviewer 2: Yes, that makes sense to me. I'm just concerned about the language, as our messaging is really important. We should also link https://github.com/pybamm-team/PyBaMM/pull/4490 as a resource for users to go through.

Author: That makes sense, I will try to make it more clear even though language is not my best suit. I already have linked the PR I am working on as a reference, I will make sure to add the scikit and nx links in the references as well.

> - An effective guide and documentation on how to use the template to make your model and distribute it.
> - A PyBaMM-based CLI to install third-party models and parameter sets on demand.

Reviewer: ```suggestion
- A guide and documentation on building and distributing custom PyBaMM models.
- A PyBaMM-based CLI to install third-party models and parameter sets on demand.
```

> - A PyBaMM-based CLI to install third-party models and parameter sets on demand.

Reviewer: Can we add the stretch goal here?

### Desired Skills

> - Experience with Python packaging and distribution for pure-Python packages and modern Python projects will be beneficial but is not required

Reviewer 2: I'd say packaging plays a bit less of a role here than we're expecting it to, as a dispatching API would be more on the programming side, with a few bits of packaging here and there.

Author: Maybe we could refactor it to versionings, tags and releases because I think it will play a huge role in setting compatibilty constraints.

> - Experience with Python packaging and distribution for pure-Python packages and modern Python projects will be beneficial but is not required
> - Familiarity with `cookiecutter/copier` templates and their templating engines
> - Basics of Git version control, CI/CD pipelines, and DevOps for Python projects

Reviewer: ```suggestion
- Python programming experience, Git version control, and GitHub workflow for open-source projects
- Experience with building and distributing pure-Python packages is beneficial but not required
- Familiarity with `cookiecutter/copier` templates and their templating engines
- An interest in scientific computing and battery modeling
```

### Difficulty

> **Medium**. This project is suitable for a 175-hour duration.

Reviewer 2: Question for all: should it be 175 or 350?

Reviewer: Sounds like a 350h project

Author: I changed it to 350 hours as we have a lot of expected outcomes after the review/

> ### Potential mentors
> 
> - [Santhosh Sundaram](https://github.com/santacodes)

Reviewer 2: ```suggestion
### Potential mentors

- [Santhosh Sundaram](https://github.com/santacodes)
- [Valentin Sulzer](https://sites.google.com/view/valentinsulzer)
- [Agriya Khetarpal](https://github.com/agriyakhetarpal)
```

Adding @valentinsulzer and myself here as secondaries 😄

@rtimms, @brosaplanella, @arjxn-py, @Saransh-cpp – feel free to hop along if you like! :)

The student will have an opportunity to learn software engineering with Python, including adding features, writing tests, writing user-facing documentation, designing examples, and building CI/CD pipelines. Additionally, their work will be a significant open-source contribution to scientific computing and and expanding battery modeling research as it will create an easy to use PyBaMM development ecosystem that any researcher could adopt, modify and distribute.

### Expected Outcomes

> - A dispatch API that could pass parameters/arguments from PyBaMM to third-party models for seamless integration. The dispatch API would also serve as a mechanism for setting constraints and compatibility checks to verify the parameters before passing them to the loaded model. The constraints could therefore be set in the model registries itself to centralise them in a single place.

Reviewer: The model registries part should be explained in more detail in a new point.

- Refactored entry points API to accommodate the dispatch API. The entry point would also serve as a loader for these third-party models that are not contained within PyBaMM, collectively, the entry points API and the dispatch API will form a framework for integrating, packaging, and distributing battery models amongst PyBaMM users.

> - An effective user experience for using third-party PyBaMM models.

Reviewer: Can we include the outcomes expected for the copier template in this point?

The student will have an opportunity to learn software engineering with Python, including adding features, writing tests, writing user-facing documentation, designing examples, and building CI/CD pipelines. Additionally, their work will be a significant open-source contribution to scientific computing and and expanding battery modeling research as it will create an easy to use PyBaMM development ecosystem that any researcher could adopt, modify and distribute.

### Expected Outcomes

- A dispatch API that could pass parameters/arguments from PyBaMM to third-party models for seamless integration. The dispatch API would also serve as a mechanism for setting constraints and compatibility checks to verify the parameters before passing them to the loaded model.

> - A registry for the models in which the constraints for a model could therefore be set in the model registries itself to centralise them in a single place. A registry could be a JSON or a configuration file defining information about a model which could be pre-loaded for checks before loading the model itself.

Reviewer: ```suggestion
- A centralized model registry where constraints for different models can be set up. The constraints in the registry can be defined as JSON or through a configuration file defining information about the model. These constraints should be pre-loaded for checks before loading the model itself.
```

- Refactored entry points API to accommodate the dispatch API. The entry point would also serve as a loader for these third-party models that are not contained within PyBaMM, collectively, the entry points API and the dispatch API will form a framework for integrating, packaging, and distributing battery models amongst PyBaMM users.

> - Feature extended copier template for incorporating features of dispatch API and simplifying the current setup processes for writing custom models and basic PyBaMM based project generation.

Reviewer: ```suggestion
- Updates to the copier template, making it compatible with the new dispatch API, simplifying the setup processes for writing custom models, and better project generation support.
```

- An effective user experience for using third-party PyBaMM models.
- A detailed and easy to read guide and documentation on building and distributing custom PyBaMM models.
- A caching and lazy loading mechanism for the entry points for efficiency.

### Desired Skills

- Python programming experience, Git version control, and GitHub workflow for open-source projects
- Experience with implementing entry points, building and distributing pure-Python packages is beneficial but not required
- Familiarity with `cookiecutter/copier` templates and their templating engines
- An interest in scientific computing and battery modeling

### Difficulty

**Medium**. This project is suitable for a 350-hour duration.

### Potential mentors

- [Santhosh Sundaram](https://github.com/santacodes)
- [Valentin Sulzer](https://sites.google.com/view/valentinsulzer)
- [Agriya Khetarpal](https://github.com/agriyakhetarpal)

### References for custom models and entry points API in PyBaMM
- https://docs.pybamm.org/en/v22.8/tutorials/add-model.html
- https://train.rse.ox.ac.uk/material/HPCu/libraries/pybamm

> - [WIP] https://github.com/pybamm-team/PyBaMM/pull/4490

Author: Yes I will make sure to add them as well!

- Refactored entry points API to accommodate the dispatch API. The entry point would also serve as a loader for these third-party models that are not contained within PyBaMM, collectively, the entry points API and the dispatch API will form a framework for integrating, packaging, and distributing battery models amongst PyBaMM users.
- Updates to the copier template, making it compatible with the new dispatch API, simplifying the setup processes for writing custom models, and better project generation support.
- An effective user experience for using third-party PyBaMM models.
- A detailed and easy to read guide and documentation on building and distributing custom PyBaMM models.
- A caching and lazy loading mechanism for the entry points for efficiency.

### Desired Skills

- Python programming experience, Git version control, and GitHub workflow for open-source projects
- Experience with implementing entry points, building and distributing pure-Python packages is beneficial but not required

> - Familiarity with `cookiecutter/copier` templates and their templating engines

Reviewer 2: ```suggestion
- Familiarity with `cookiecutter`/`copier` templates and their templating engines
```

Reviewer 3: Let's merge this :)
